Bump kube-prometheus-stack to 45.5.0

gdemonet opened the pull request:
Update the charts with:

```
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update
rm -rf charts/kube-prometheus-stack
helm fetch -d charts --untar prometheus-community/kube-prometheus-stack
```

Also bump a number of components:
- Prometheus to 2.42.0
- Thanos to 0.30.2
- Grafana to 9.3.8
- kiwigrid/k8s-sidecar to 1.22.3
- kube-state-metrics to 2.8.0
- node-exporter to 1.5.0
- prometheus-operator to 0.63.0

The chart.sls was re-rendered with:

```
./doit.sh codegen:chart_kube-prometheus-stack
```

Since we bumped Thanos, we also re-render its own chart (without
updating, since it did not change since the last update) with:

```
./doit.sh codegen:chart_thanos
```

Important note: Alertmanager configuration was updated by hand,
and a note was added to try and remind maintainers of doing it in the
future. This should help making the "InfoInhibitor" alert more useful.

gdemonet committed:
build,charts,salt: Bump kube-prometheus-stack to 45.5.0

Update the charts with:

```
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update
rm -rf charts/kube-prometheus-stack
helm fetch -d charts --untar prometheus-community/kube-prometheus-stack
```

Also bump a number of components:
- Prometheus to 2.42.0
- Thanos to 0.30.2
- Grafana to 9.3.8
- kiwigrid/k8s-sidecar to 1.22.3
- kube-state-metrics to 2.8.0
- node-exporter to 1.5.0
- prometheus-operator to 0.63.0

The chart.sls was re-rendered with:

```
./doit.sh codegen:chart_kube-prometheus-stack
```

Since we bumped Thanos, we also re-render its own chart (without
updating, since it did not change since the last update) with:

```
./doit.sh codegen:chart_thanos
```

Important note: Alertmanager configuration was updated by hand,
and a note was added to try and remind maintainers of doing it in the
future. This should help making the "InfoInhibitor" alert more useful.

gdemonet committed:
salt,charts: Allow defining registry separately

Some charts now expect the image registry to be defined separately from
the repository, and enforces that these values are joined with a slash.
This causes issues with our macro "build_image_name", which builds the
whole path.
We add an option to this macro to omit the registry endpoint, and make
this value available to rendered charts using the charts/render.py
script header.

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/metalk8s/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @gdemonet

* @lucieleonard

* @jbwatenbergscality

* @teddyandrieux

* @chengyanjin

TeddyAndrieux commented on charts/kube-prometheus-stack.yaml:
```diff
@@ -12,7 +12,8 @@ commonLabels:
 alertmanager:
   alertmanagerSpec:
     image:
-      repository: '__image__(alertmanager)'
+      registry: '__var__(repo.registry_endpoint)'
+      repository: '__image_no_reg__(alertmanager)'
```
Sad, but ok, yes it's needed

But cannot we set this registry only once in `global` ?

gdemonet replied:
Right will do, found out too late about this one 😇

gdemonet replied:
Argh no I can't do that because the kube-state-metrics subchart doesn't use this value correctly:

```
image: "{{ .Values.global.imageRegistry | default .Values.image.repository }}:{{ .Values.image.tag | ...
```

:sob: 

I'll submit a PR over there to have it fixed

TeddyAndrieux commented on charts/kube-prometheus-stack/README.md:
````diff
@@ -80,6 +80,174 @@ _See [helm upgrade](https://helm.sh/docs/helm/helm_upgrade/) for command documen
+kubectl apply --server-side -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/v0.59.1/example/prometheus-operator-crd/monitoring.coreos.com_thanosrulers.yaml
+```
+
+Starting from prometheus-node-exporter version 4.0.0, the `node exporter` chart is using the [Kubernetes recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/). Therefore you have to delete the daemonset before you upgrade.
````
:eyes: 

We need to handle it

gdemonet replied:
Oh yes good catch, didn't check upgrade notes 🤦

TeddyAndrieux reviewed (commented):
Really really hard to review

Just by looking at the changes on the salt side it looks good to me (except for the upgrade handling)

TeddyAndrieux commented on CHANGELOG.md:
```diff
@@ -56,13 +56,55 @@
 
+- Bump kube-prometheus-stack chart version to
+  [45.5.0](https://github.com/prometheus-community/helm-charts/releases/tag/kube-prometheus-stack-45.5.0)
+  (PR[#4017](https://github.com/scality/metalk8s/pull/4011))
```
typo: 
```suggestion
  (PR[#4017](https://github.com/scality/metalk8s/pull/4017))
```

TeddyAndrieux commented on CHANGELOG.md:
```diff
@@ -56,13 +56,55 @@
+
+- Some metrics were removed when bumping kube-prometheus-stack chart to
+  [45.5.0](https://github.com/prometheus-community/helm-charts/releases/tag/kube-prometheus-stack-45.5.0)
+  (PR[#4017](https://github.com/scality/metalk8s/pull/4011)):
```
ditto

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/bump-kube-prometheus-and-thanos` and the
destination branch `development/125.0`.

Please resolve the conflict on **the feature branch** (`improvement/bump-kube-prometheus-and-thanos`).

```sh
git fetch && \
git checkout origin/improvement/bump-kube-prometheus-and-thanos && \
git merge origin/development/125.0
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:improvement/bump-kube-prometheus-and-thanos
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @gdemonet

* @lucieleonard

* @jbwatenbergscality

* @teddyandrieux

* @chengyanjin

gdemonet commented on charts/kube-prometheus-stack/values.yaml:
```diff
@@ -897,7 +1057,31 @@ kubelet:
+      - sourceLabels: [__name__]
+        action: drop
+        regex: 'container_spec.*'
+      # Drop cgroup metrics with no pod.
+      - sourceLabels: [id, pod]
+        action: drop
+        regex: '.+;'
```
This causes the current prometheus-adapter `resourceRules.(cpu|memory).nodeQuery` to not hit anything, because we dropped the metrics for nodes (`pod="", id="/"`).

It appears the goal is to now rely on more efficient node-exporter metrics (see https://github.com/kubernetes-sigs/prometheus-adapter/pull/516 and https://github.com/prometheus-community/helm-charts/pull/2827), but we're now hitting an issue with labels: our node-exporter metrics don't have a "node" label, and that's not good when trying to map to `/metrics.k8s.io/v1beta1/nodes`.

We will need to fix this label issue (TBH, would be much simpler to explore and query, e.g. from our UI), but I'm not sure how involved this will be. For now, I'm considering a temporary workaround by keeping these metrics around, with a follow-up ticket.

gdemonet replied:
Here's the follow-up https://github.com/scality/metalk8s/issues/4018

gdemonet replied:
And the workaround: https://github.com/scality/metalk8s/pull/4017/commits/97ce7c66bedb148d381a9f1011b4afdfc70b6106

gdemonet committed:
charts,salt: Reinstate node-level cAdvisor metrics

This changes the default configuration from kube-prometheus-stack since
we still use these metrics in prometheus-adapter.
Ideally, we would rather let prometheus-adapter consume node-exporter
metrics, but this requires #4018 to be fixed first.

gdemonet committed:
tests/post: Wait longer for new Grafana dashboard

Had a flaky on this (failed on single-node but multi-nodes succeeded),
let's wait a bit longer.

gdemonet commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @gdemonet

* @lucieleonard

* @jbwatenbergscality

* @teddyandrieux

* @chengyanjin










*The following options are set:* **approve**

TeddyAndrieux commented on salt/metalk8s/addons/prometheus-operator/pre-upgrade.sls:
```diff
@@ -0,0 +1,29 @@
+{#- Node-exporter statefulset now uses recommended labels, which are immutable,
+    so we need to delete it first.
+    Can be removed in development/126.0 #}
+
+{%- set node_exporter_sts = salt.metalk8s_kubernetes.get_object(
+    kind="StatefulSet",
+    apiVersion="apps/v1",
+    name="prometheus-operator-prometheus-node-exporter",
+    namespace="metalk8s-monitoring"
+) %}
+
+{%- if node_exporter_sts and salt.pkg.version_cmp(
+        node_exporter_sts["metadata"]["labels"]["metalk8s.scality.com/version"],
+        "125.0.0"
+    ) == -1 %}
+
+Delete old node-exporter StatefulSet:
+    metalk8s_kubernetes.object_absent:
+        - apiVersion: apps/v1
+        - kind: StatefulSet
+        - name: prometheus-operator-prometheus-node-exporter
+        - namespace: metalk8s-monitoring
+
+{%- else %}
+
+Nothing to do before upgrading for prometheus-operator:
+  test.nop: []
+
+{%- endif %}
```
Usually I prefer to do it directly in the `init.sls` of the deployed states (so that we only remove it when we are actually deploying the new one)

And we also need to do it for downgrade btw
(https://github.com/scality/metalk8s/commit/f30023586692d4672dc5d2a4c0d12c7cad987bab)

gdemonet replied:
Ah yes, downgrade 😬 On it

gdemonet replied:
Done

TeddyAndrieux reviewed (approved):
:shipit:

TeddyAndrieux commented on salt/metalk8s/addons/prometheus-operator/deployed/init.sls:
```diff
@@ -8,3 +8,28 @@ include:
+        - apiVersion: apps/v1
+        - kind: StatefulSet
+        - name: prometheus-operator-prometheus-node-exporter
+        - namespace: metalk8s-monitoring
```
Oh we should `wait` for deletion here and `require_in` the `sls` otherwise it will fail on the first upgrade run

gdemonet replied:
Good catch, eagle eyes :eagle: !

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @gdemonet

* @lucieleonard

* @jbwatenbergscality

* @teddyandrieux

* @chengyanjin





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/b6e5ba9fd5456dfe9e39a1b2973f21cfceea8ca4) did not succeed in branch improvement/bump-kube-prometheus-and-thanos.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/125.0`



The following branches will **NOT** be impacted:

* `development/123.0`
* `development/124.0`
* `development/124.1`
* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/125.0`



The following branches have **NOT** changed:

* `development/123.0`
* `development/124.0`
* `development/124.1`
* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.11`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye gdemonet.
